banzai/settings.py

```diff
@@ -7,122 +7,77 @@
+                  'banzai.gain.GainNormalizer',
+                  'banzai.mosaic.MosaicCreator',
+                  'banzai.trim.Trimmer',
+                  'banzai.bias.BiasSubtractor',
+                  'banzai.dark.DarkSubtractor',
+                  'banzai.flats.FlatDivider',
+                  'banzai.qc.PatternNoiseDetector',
+                  'banzai.photometry.SourceDetector',
+                  'banzai.astrometry.WCSSolver',
+                  'banzai.qc.pointing.PointingTest']
+
+CALIBRATION_MIN_FRAMES = {'BIAS': 5,
+                          'DARK': 5,
+                          'SKYFLAT': 5}
+
+CALIBRATION_SET_CRITERIA = {'BIAS': ['ccdsum'],
+                            'DARK': ['ccdsum'],
+                            'SKYFLAT': ['ccdsum', 'filter']}
+
+LAST_STAGE = {'BIAS': 'banzai.trim.Trimmer', 'DARK': 'banzai.bias.BiasSubtractor', 'SKYFLAT': 'banzai.dark.DarkSubtractor',
+              'SINISTRO': 'banzai.mosaic.MosaicCreator', 'STANDARD': None, 'EXPOSE': None}
+
+EXTRA_STAGES = {'BIAS': ['banzai.bias.BiasMasterLevelSubtractor', 'banzai.bias.BiasComparer'],
+                'DARK': ['banzai.dark.DarkNormalizer', 'banzai.dark.DarkComparer'],
+                'SKYFLAT': ['banzai.flats.FlatNormalizer', 'banzai.qc.PatternNoiseDetector', 'banzai.flats.FlatComparer'],
+                'STANDARD': None,
+                'EXPOSE': None}
+
+CALIBRATION_STACKER_STAGE = {'BIAS': 'banzai.bias.BiasMaker',
+                             'DARK': 'banzai.dark.DarkMaker',
+                             'SKYFLAT': 'banzai.flats.FlatMaker'}
+
+CALIBRATION_IMAGE_TYPES = ['BIAS', 'DARK', 'SKYFLAT']
+
+SINISTRO_IMAGE_TYPES = ['BIAS', 'DARK', 'SKYFLAT', 'EXPOSE', 'STANDARD', 'TRAILED', 'EXPERIMENTAL']
+
+
+def make_calibration_filename_function(calibration_type, attribute_filename_functions, telescope_filename_function):
+    def get_calibration_filename(image):
+        name_components = {'site': image.site, 'telescop': telescope_filename_function(image),
+                           'camera': image.header.get('INSTRUME', ''), 'epoch': image.epoch,
+                           'cal_type': calibration_type.lower()}
+        cal_file = '{site}{telescop}-{camera}-{epoch}-{cal_type}'.format(**name_components)
+        for filename_function in attribute_filename_functions:
+            cal_file += '-{}'.format(filename_function(image))
+        cal_file += '.fits'
+        return cal_file
+    return get_calibration_filename
+
+
+CALIBRATION_FILENAME_FUNCTIONS = {'BIAS': make_calibration_filename_function('BIAS', [ccdsum_to_filename], telescope_to_filename),
+                                  'DARK': make_calibration_filename_function('DARK', [ccdsum_to_filename], telescope_to_filename),
+                                  'SKYFLAT': make_calibration_filename_function('SKYFLAT', [ccdsum_to_filename, filter_to_filename],
+                                                                                telescope_to_filename)}
```

Comment:
It seems that the two big changes to `settings.py` are that 1) the required settings parameters are no longer "enforced" since they are no longer properties, 2) there is no longer a distinction between general vs. imaging settings. 

Given point 2, how will the settings be implemented in BANZAI NRES? Have you thought about adding some kind of `settings_type` variable to `realtime_context`, which would specify the settings import type? Then the two sets of settings could remain distinct.

Reply:
After discussion I realized that this is not really possible so don't worry about this comment.